core/app/components/FormBuilder/FormBuilder.tsx

```diff
@@ -107,13 +114,98 @@ type Props = {
+ */
+const preparePayload = ({
+	formValues,
+	defaultValues,
+}: {
+	defaultValues: FormBuilderSchema;
+	formValues: FormBuilderSchema;
+}) => {
+	const { upserts, deletes, relatedPubTypes, deletedRelatedPubTypes } =
+		formValues.elements.reduce<{
+			upserts: NewFormElements[];
+			deletes: FormElementsId[];
+			relatedPubTypes: NewFormElementToPubType[];
+			deletedRelatedPubTypes: FormElementsId[];
+		}>(
+			(acc, element, index) => {
+				if (element.deleted) {
+					if (element.elementId) {
+						acc.deletes.push(element.elementId);
+					}
+				} else if (!element.elementId) {
+					// Newly created elements have no elementId, so generate an id to use
+					const id = crypto.randomUUID() as FormElementsId;
+					acc.upserts.push(
+						formElementsInitializerSchema.parse({
+							formId: formValues.formId,
+							...element,
+							id,
+						})
+					);
+					if (element.relatedPubTypes) {
+						for (const pubTypeId of element.relatedPubTypes) {
+							acc.relatedPubTypes.push({ A: id, B: pubTypeId });
+						}
+					}
+				} else if (element.updated) {
+					// check whether the element is reeeaally updated minus the updated field
+					const { updated: _, id: _id, ...elementWithoutUpdated } = element;
+					const { updated, id, ...rest } =
+						defaultValues.elements.find((e) => e.elementId === element.elementId) ?? {};
+
+					const defaultElement = rest as Omit<FormElementData, "updated" | "id">;
+
+					if (JSON.stringify(defaultElement) === JSON.stringify(elementWithoutUpdated)) {
+						return acc;
+					}
+
+					acc.upserts.push(
+						formElementsInitializerSchema.parse({
+							...element,
+							formId: formValues.formId,
+							id: element.elementId,
+						})
+					); // TODO: only update changed columns
+					if (element.relatedPubTypes) {
+						// If we are updating to an empty array and there were related pub types before, we should clear out all related pub types
+						if (
+							element.relatedPubTypes.length === 0 &&
+							defaultElement.relatedPubTypes?.length
+						) {
+							acc.deletedRelatedPubTypes.push(element.elementId);
+						} else {
+							for (const pubTypeId of element.relatedPubTypes) {
+								acc.relatedPubTypes.push({ A: element.elementId, B: pubTypeId });
+							}
+						}
+					}
+				}
+				return acc;
+			},
+			{ upserts: [], deletes: [], relatedPubTypes: [], deletedRelatedPubTypes: [] }
+		);
+
+	const access = formValues.access !== defaultValues.access ? formValues.access : undefined;
+
+	return {
+		formId: formValues.formId,
+		upserts,
+		deletes,
+		access,
+		relatedPubTypes,
+		deletedRelatedPubTypes,
+	};
+};
+
 export function FormBuilder({ pubForm, id, stages }: Props) {
-	const router = useRouter();
-	const pathname = usePathname();
-	const params = useSearchParams();
-	const form = useForm<FormBuilderSchema>({
-		resolver: zodResolver(formBuilderSchema),
-		values: {
+	const [isChanged, setIsChanged] = useIsChanged();
```

Comment:
this is a small custom `nuqs` hook reads from/updates the `isChanged` query param. using nuxt we can do this without a round trip to the server, which we want.

https://github.com/pubpub/platform/blob/415de673452e7952872c1d341ebfc01f3da8054c/core/app/components/FormBuilder/useIsChanged.tsx#L1-L13